### DISCUSSION TITLE
Conventions FindClrType to include entity type information?

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#15593] [Ideas]
The following `DocumentConventions` method is used to call the `FindClrType` convention:
https://github.com/ravendb/ravendb/blob/b4add7c253da20a5bf1a630e1910501189d0a0f4/src/Raven.Client/Documents/Conventions/DocumentConventions.cs#L1046-L1049

Do you think it would make sense to include the expected entity type alongside the document id in `FindClrType` _(the methods depending on FindClrType seem to have this info)_?

Asking as that would help determining the correct type as we could filter by the expected entity type.

If we do not want to depend on `@metadata`, that might or might not have the information we need, the only way we can infer the type is by analyzing the provided document id. In case that is a meaningless GUID we are out of luck.

**Comment by garayx** [comment]
I guess you can put the type name as a property in the metadata of the document (or the document itself) and then override the `GetClrType` convention , 


                Conventions = new DocumentConventions
                {
                    FindClrType = (id, o) => {
                        if (o.TryGetMember("MyType", out object obj) && obj is string type)
                            return type;

                        return null;
                    }
                }

**Comment by ayende** [comment]
What is it that you are trying to do?
This is there if you need to have some custom logic dictating this value.
For example, you have a Python app and a C# one collaborating on the same data, and you need to maintain similar inheritance hierarchies.

**Reply by mcrio** [reply]
I am just evaluating some scenarios, like someone might add an item to a collection from the Studio without the Clr-Type metadata. I could lookup the ID or collection name and try to find the type that way, but if the code already knows what it expects it'll be useful to provide that info to `FindClrType`.

**Reply by ayende** [reply]
In almost always cases, you do something like `Load<Type>(...)` which means that the client will use the `<Type>` to serialize the instance if the `Raven-Clr-Type` isn't there.

**Reply by mcrio** [reply]
The problem is with derived classes where you specify the base class during query/load.

**Reply by ayende** [reply]
So that is a reason to use `FindClrType` , yes

**Reply by mcrio** [reply]
Let's assume we don't have the CLR type in the metadata.
Using `FindClrType` the only clue about the type of that document would be the metadata `@collection` value. Based on that we could examine the structure and figure out the proper derived type. As I see we seem to already have the expected **base** type that is given during query or load, so why not passing down to `FindClrType` as it'd help figure out the proper derived type.

**Reply by ayende** [reply]
`FindClrType` may be called on many scenarios, including `Load<object>()`, you cannot rely on the context clues for that.
You need to be able to derive that from the document itself.

**Reply by mcrio** [reply]
Ok thanks.
